### PR TITLE
test/{fs,cephfs}: Get libcephfs and cephfs to compile with FreeBSD

### PR DESCRIPTION
**Pull request by wjwithagen**
FreeBSD does not have features.h
FreeBSD dirent{} does not have d_off, so set the offset pointer to 0
FreeBSD does not have sys/xattr.h
    But only the XATTR_CREATE define looks to be missing

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`

</details>

### CODE REVIEW

File: src/test/libcephfs/flock.cc [hunk 1]
```diff
@@ -175,27 +178,27 @@ static void thread_ConcurrentLocking(str_ConcurrentLocking& s) {
   ASSERT_GE(fd, 0); 
 
   ASSERT_EQ(-EWOULDBLOCK,
-	    ceph_flock(cmount, fd, LOCK_EX | LOCK_NB, pthread_self()));
+	    ceph_flock(cmount, fd, LOCK_EX | LOCK_NB, (uint64_t)pthread_self()));
```
Comment (tchaikov) [inline src/test/libcephfs/flock.cc]: assuming `pthread_self()` returns `uint64_t` is wrong. but this is also incorrect. i'd suggest add 
```c++
uint64_t get_owner() {
  auto me = pthread_self();
  static_assert(std::is_convertible_v<decltype(me), uint64_t> ||
                std::is_pointer_v<decltype(me)>,
                "we need to use pthread_self() for the owner parameter");
  return reinterpret_cast<uint64_t>(me);
}
```
to document this assumption, and s/pthread_self/get_owner/ in the tests.

File: src/test/libcephfs/flock.cc [hunk 2]
```diff
@@ -26,15 +26,18 @@
...
-
 #include <stdlib.h>
 #include <semaphore.h>
 #include <time.h>
 #include <sys/mman.h>
 
 #ifdef __linux__
 #include <limits.h>
+#include <sys/xattr.h>
+#endif
```
[the remark this answers is not in the document]
Comment (wjwithagen) [reply src/test/libcephfs/flock.cc]: @tchaikov 
Technically that is not true....
There is somethink like: Debian GNU/kFreeBSD
And I've tried fixing some RocksDB conditionaling, and that is not easy...
But I get your point.

[the remark this answers is not in the document]
Comment (tchaikov) [reply src/test/libcephfs/flock.cc]: @wjwithagen i think you are taking GNU as Linux. and i am not sure how this is related.

File: src/test/libcephfs/flock.cc (hunk 1 above)
Comment (tchaikov) [reply src/test/libcephfs/flock.cc]: how about putting this code snippet in this test.

Comment (wjwithagen) [reply src/test/libcephfs/flock.cc]: @tchaikov 
For now I have it in `compat.{h,cc}`.
Since it will allow to "fix" the other usuage.

Comment (tchaikov) [reply src/test/libcephfs/flock.cc]: i don't it's **right** to do so. as it's not related to compatibility, actually it defeats the purpose of compatibility. per the POSIX standard, `pthread_t` is an opaque type, which could be a pointer, an integer or a structure. it's us who are using the fact that it's an `unsigned long` in glibc.

Comment (wjwithagen) [reply src/test/libcephfs/flock.cc]: @tchaikov 
That is true....
I'll throw it into `src/test/libcephfs/`

File: src/test/libcephfs/ceph_pthread_self.cc [hunk 3]
```diff
@@ -0,0 +1,26 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2011 New Dream Network
```
Comment (batrick) [inline src/test/libcephfs/ceph_pthread_self.cc]: I think you can omit the copyright header. @tchaikov ?

File: src/test/libcephfs/ceph_pthread_self.cc [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+#include <type_traits>
+
+uint64_t ceph_pthread_self() {
+  auto me = pthread_self();
+  static_assert(std::is_convertible_v<decltype(me), uint64_t> ||
+                std::is_pointer_v<decltype(me)>,
+                "we need to use pthread_self() for the owner parameter");
+  return reinterpret_cast<uint64_t>(me);
+}
```
Comment (jtlayton) [inline src/test/libcephfs/ceph_pthread_self.cc]: pthread_self returns pthread_t on both Linux and BSD, AFAICT. Why is this needed?

File: src/test/libcephfs/ceph_pthread_self.cc (hunk 3 above)
Comment (tchaikov) [reply src/test/libcephfs/ceph_pthread_self.cc]: yeah. please drop the boilerplate header. and i don't think it's necessary to have another .cc file for this.

Comment (wjwithagen) [reply src/test/libcephfs/ceph_pthread_self.cc]: Yeah, I'm a stickler from the old age where code went into .c{c} files....
Will fix.

**Comment by jbeich** [comment]
> FreeBSD `struct dirent` does not have `d_off`, so set the offset pointer to 0

[It does](https://svnweb.freebsd.org/base/head/sys/sys/dirent.h?r1=318736&r2=318735&pathrev=318736) since FreeBSD 12.0 switched to 64-bit inodes.

File: src/test/libcephfs/ceph_pthread_self.cc (hunk 4 above)
Comment (wjwithagen) [reply src/test/libcephfs/ceph_pthread_self.cc]: @jtlayton 
Because the gtest code wants something that resembles a `uint64_t` and  libc shipped with FreeBSD has an opaque `pthread_t`. So conversion needs some helping to get it done.
And this snippet was suggested by @tchaikov, instead of just a raw casting prefix.
And I guess that gcc and glibc are more flexible "to do the right thing"

Comment (jtlayton) [reply src/test/libcephfs/ceph_pthread_self.cc]: Yuck -- this seems like a bug in gtest, fwiw. In the meantime though this seems like a reasonable workaround. Could you explain this better in the changelog? At some point gtest may stop requiring this, and we may be able to revert this change.

**Comment by tchaikov** [comment]
i am removing clang from the title of this PR. see https://github.com/ceph/ceph/pull/30505#discussion_r326866511

File: src/test/libcephfs/ceph_pthread_self.h [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+// vim: ts=8 sw=2 smarttab
+
+#ifndef CEPH_TEST_LIBCEPHFS_PTHREAD_SELF
+#define CEPH_TEST_LIBCEPHFS_PTHREAD_SELF
+
+#include <pthread.h>
+
+#include <type_traits>
+
+static uint64_t ceph_pthread_self() {
```
Comment (jtlayton) [inline src/test/libcephfs/ceph_pthread_self.h]: It would be nice to have a comment block above this that explains why it's needed for posterity. Eventually, maybe gtest will handle this in a cleaner fashion and we'll be able to revert to just using pthread_self internally rather than this wrapper.

Comment (wjwithagen) [reply src/test/libcephfs/ceph_pthread_self.h]: @jtlayton 
I can still do that.
Then the question becomes, how do we find out when `gtest` has been cleaned up?

**Comment by tchaikov** [comment]
https://shaman.ceph.com/builds/ceph/wip-kefu-testing-2019-09-30-1336/2c0b2169f6303e11ab6bed736272e5e2d23ab027/

@batrick do we need to run this through cephfs suite?

**Comment by batrick** [comment]
@tchaikov I think it can't hurt. I'll queue it up for the next run.

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2019-10-09_15:12:02-rados-wip-kefu-testing-2019-10-09-1243-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2019-10-14_06:20:56-rados-wip-kefu-testing-2019-10-14-1111-distro-basic-smithi/

failures tracked by 

- https://tracker.ceph.com/issues/42301
- https://tracker.ceph.com/issues/40026
